cmd/titus-storage/ebs.go

```diff
@@ -0,0 +1,272 @@
+			log.Print(err)
+		}
+		log.Print(response)
+
+		// TODO: Check and sleep
+	}
+	return true
+}
+
+func randDriveNamePicker() (string, error) {
+	ctr := 0
+	deviceName := "/dev/xvd"
+	runes := []rune("fghijklmnopqrstuvwxyz")
+	for {
+		if ctr >= len(runes) {
+			return "", fmt.Errorf("Ran out of drive names")
+		}
+		if !doesDriveExist(deviceName + string(runes[ctr])) {
+			break
+		}
+		ctr++
+	}
+	return deviceName + string(runes[ctr]), nil
+}
+
+func doesDriveExist(driveName string) bool {
+	if _, err := os.Stat(driveName); os.IsNotExist(err) {
+		for _, file := range listNvmeBlockDevices() {
+			if d, _ := ebsnvme.ScanDevice(file); d.Name == driveName {
+				return true
+			}
+		}
+		return false
+	}
+	return true
+}
+
+func waitForDeviceToDisappear(driveName string, maxAttempts int) bool {
+	var attempts int
+	for doesDriveExist(driveName) {
+		time.Sleep(2 * time.Second)
+		attempts++
+		if attempts >= maxAttempts {
+			log.Printf("Waited the max %d seconds for %s to go, but it is still here!", attempts*2, driveName)
+			return false
+		}
+	}
+	log.Printf("Waited for %s to disappear, and it is now gone after %d seconds", driveName, attempts*2)
+	return true
+}
+
+func doesDriveExistWithTimeout(driveName string, maxAttempts int) bool {
+	var attempts int
+	for !doesDriveExist(driveName) {
+		time.Sleep(2 * time.Second)
+		attempts++
+		if attempts >= maxAttempts {
+			return false
+		}
+	}
+	return true
+}
+
+// GetActualBlockDeviceName returns the actual name of the block device seen
+// within the instance, like /dev/nvme1
+// This is often different from the ec2 device name, which is usually
+// something like /dev/sdx
+func GetActualBlockDeviceName(name string) (string, error) {
+	for _, device := range listNvmeBlockDevices() {
+		if d, _ := ebsnvme.ScanDevice(device); d.Name == name {
+			return device, nil
+		}
+	}
+	if _, err := os.Stat(name); os.IsNotExist(err) {
+		return "", err
+	}
+	return name, nil
+}
+
+func listNvmeBlockDevices() (devices []string) {
+	re := regexp.MustCompile(`(^\/dev\/nvme[0-9]+n1$)`)
+	f, _ := filepath.Glob("/dev/nvme*")
+	for _, d := range f {
+		if re.Match([]byte(d)) {
+			devices = append(devices, d)
+		}
+	}
+	return
+}
+
+// getDeviceName scans the Nvme on disk and looks at the
+// serial numbers of the devices and returns the device name
+// (/dev/nmve1). Behaves like a hash get, returning key, ok.
+// returning "", false if unfound.
+func getDeviceName(ebsVolumeID string) (string, bool) {
+	deviceName := ""
+	for _, device := range listNvmeBlockDevices() {
+		if d, _ := ebsnvme.ScanDevice(device); d.VolumeID == ebsVolumeID {
+			deviceName = d.Name
+		}
+	}
+	return deviceName, deviceName != ""
+}
+
+func attachEBS(ebsVolumeID string, ec2Client *ec2.EC2, instanceID string) (string, error) {
+	deviceName, ok := getDeviceName(ebsVolumeID)
+	if ok {
+		log.Printf("EBS Volume %s already attached as %s", ebsVolumeID, deviceName)
+		return deviceName, nil
+	}
+	deviceName, err := randDriveNamePicker()
```

Comment:
Is this racey?  If you have 2 different attaches happening at the same time, and they're going through the list in the same order (which it looks like they are), then it's possible that they both pick the same drive name, and the slower one fails, right?  This gets much more likely if the attach takes many seconds.  I'm wondering if this function should retry on an error from trying to attach a drive name that already exists.

(I know this is a Draft PR, so it may well be that you're planning on fixing this later, but figured I'd ask.)

Reply:
Hmm. In my bash version I `flockd` everything so that there was never a chance for races on the single host.

But I think it would be simpler to just fail and let titus relaunch it with its own backoff.

Reply:
We do have `FSLocker` in titus-executor, if you want to use it.

Reply:
Hmm, this is all pretty silly, because this device name (/dev/xvda) isn't actually used in practice with nitro/nvme.